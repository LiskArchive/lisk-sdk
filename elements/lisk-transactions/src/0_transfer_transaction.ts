/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import * as BigNum from '@liskhq/bignum';
import { getAddressFromPublicKey } from '@liskhq/lisk-cryptography';
import {
	BaseTransaction,
	StateStore,
	StateStorePrepare,
} from './base_transaction';
import { MAX_TRANSACTION_AMOUNT, TRANSFER_FEE } from './constants';
import { convertToAssetError, TransactionError } from './errors';
import { TransactionJSON } from './transaction_types';
import {
	validateAddress,
	validateTransferAmount,
	validator,
	verifyAmountBalance,
	verifyBalance,
} from './utils';

const TRANSACTION_TRANSFER_TYPE = 0;

export interface TransferAsset {
	readonly data: string;
}

export const transferAssetFormatSchema = {
	type: 'object',
	properties: {
		data: {
			type: 'string',
			format: 'noNullByte',
			maxLength: 64,
		},
	},
};

export class TransferTransaction extends BaseTransaction {
	public readonly asset: TransferAsset;

	public constructor(rawTransaction: unknown) {
		super(rawTransaction);
		const tx = (typeof rawTransaction === 'object' && rawTransaction !== null
			? rawTransaction
			: {}) as Partial<TransactionJSON>;
		// Initializes to empty object if it doesn't exist
		this.asset = (tx.asset || {}) as TransferAsset;
	}

	protected assetToBytes(): Buffer {
		const { data } = this.asset;

		return data ? Buffer.from(data, 'utf8') : Buffer.alloc(0);
	}

	public assetToJSON(): TransferAsset {
		return this.asset;
	}

	public async prepare(store: StateStorePrepare): Promise<void> {
		await store.account.cache([
			{
				address: this.senderId,
			},
			{
				address: this.recipientId,
			},
		]);
	}

	// tslint:disable-next-line prefer-function-over-method
	protected verifyAgainstTransactions(
		_: ReadonlyArray<TransactionJSON>,
	): ReadonlyArray<TransactionError> {
		return [];
	}

	protected validateAsset(): ReadonlyArray<TransactionError> {
		validator.validate(transferAssetFormatSchema, this.asset);
		const errors = convertToAssetError(
			this.id,
			validator.errors,
		) as TransactionError[];

		if (this.type !== TRANSACTION_TRANSFER_TYPE) {
			errors.push(
				new TransactionError(
					'Invalid type',
					this.id,
					'.type',
					this.type,
					TRANSACTION_TRANSFER_TYPE,
				),
			);
		}

		if (!validateTransferAmount(this.amount.toString())) {
			errors.push(
				new TransactionError(
					'Amount must be a valid number in string format.',
					this.id,
					'.amount',
					this.amount.toString(),
				),
			);
		}

		if (!this.fee.eq(TRANSFER_FEE)) {
			errors.push(
				new TransactionError(
					`Fee must be equal to ${TRANSFER_FEE}`,
					this.id,
					'.fee',
					this.fee.toString(),
					TRANSFER_FEE,
				),
			);
		}

		if (!this.recipientId) {
			errors.push(
				new TransactionError(
					'`recipientId` must be provided.',
					this.id,
					'.recipientId',
				),
			);
		}

		try {
			validateAddress(this.recipientId);
		} catch (error) {
			errors.push(
				new TransactionError(
					error.message,
					this.id,
					'.recipientId',
					this.recipientId,
				),
			);
		}

		if (this.recipientPublicKey) {
			const calculatedAddress = getAddressFromPublicKey(
				this.recipientPublicKey,
			);
			if (this.recipientId !== calculatedAddress) {
				errors.push(
					new TransactionError(
						'recipientId does not match recipientPublicKey.',
						this.id,
						'.recipientId',
						this.recipientId,
						calculatedAddress,
					),
				);
			}
		}

		return errors;
	}

	protected applyAsset(store: StateStore): ReadonlyArray<TransactionError> {
		const errors: TransactionError[] = [];
		const sender = store.account.get(this.senderId);

		const balanceError = verifyAmountBalance(
			this.id,
			sender,
			this.amount,
			this.fee,
		);
		if (balanceError) {
			errors.push(balanceError);
		}

		const updatedSenderBalance = new BigNum(sender.balance).sub(this.amount);

		const updatedSender = {
			...sender,
			balance: updatedSenderBalance.toString(),
		};
		store.account.set(updatedSender.address, updatedSender);
		const recipient = store.account.getOrDefault(this.recipientId);

		const updatedRecipientBalance = new BigNum(recipient.balance).add(
			this.amount,
		);

		if (updatedRecipientBalance.gt(MAX_TRANSACTION_AMOUNT)) {
			errors.push(
				new TransactionError(
					'Invalid amount',
					this.id,
					'.amount',
					this.amount.toString(),
				),
			);
		}

		const updatedRecipient = {
			...recipient,
			balance: updatedRecipientBalance.toString(),
		};
		store.account.set(updatedRecipient.address, updatedRecipient);

		return errors;
	}

	protected undoAsset(store: StateStore): ReadonlyArray<TransactionError> {
		const errors: TransactionError[] = [];
		const sender = store.account.get(this.senderId);
		const updatedSenderBalance = new BigNum(sender.balance).add(this.amount);

		if (updatedSenderBalance.gt(MAX_TRANSACTION_AMOUNT)) {
			errors.push(
				new TransactionError(
					'Invalid amount',
					this.id,
					'.amount',
					this.amount.toString(),
				),
			);
		}

		const updatedSender = {
			...sender,
			balance: updatedSenderBalance.toString(),
		};
		store.account.set(updatedSender.address, updatedSender);
		const recipient = store.account.getOrDefault(this.recipientId);

		const balanceError = verifyBalance(this.id, recipient, this.amount);

		if (balanceError) {
			errors.push(balanceError);
		}

		const updatedRecipientBalance = new BigNum(recipient.balance).sub(
			this.amount,
		);

		const updatedRecipient = {
			...recipient,
			balance: updatedRecipientBalance.toString(),
		};

		store.account.set(updatedRecipient.address, updatedRecipient);

		return errors;
	}
}
