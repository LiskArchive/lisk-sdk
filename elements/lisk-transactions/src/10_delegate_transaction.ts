/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { validator } from '@liskhq/lisk-validator';

import {
	BaseTransaction,
	StateStore,
	StateStorePrepare,
} from './base_transaction';
import { DELEGATE_FEE } from './constants';
import { convertToAssetError, TransactionError } from './errors';
import { Account, TransactionJSON } from './transaction_types';

export interface DelegateAsset {
	readonly username: string;
}

export const delegateAssetFormatSchema = {
	type: 'object',
	required: ['username'],
	properties: {
		username: {
			type: 'string',
			minLength: 1,
			maxLength: 20,
			format: 'username',
		},
	},
};

export class DelegateTransaction extends BaseTransaction {
	public readonly asset: DelegateAsset;
	public readonly containsUniqueData: boolean;
	public static TYPE = 10;
	public static FEE = DELEGATE_FEE.toString();

	public constructor(rawTransaction: unknown) {
		super(rawTransaction);
		const tx = (typeof rawTransaction === 'object' && rawTransaction !== null
			? rawTransaction
			: {}) as Partial<TransactionJSON>;
		this.asset = (tx.asset || { delegate: {} }) as DelegateAsset;
		this.containsUniqueData = true;
	}

	protected assetToBytes(): Buffer {
		const { username } = this.asset;

		return Buffer.from(username, 'utf8');
	}

	public async prepare(store: StateStorePrepare): Promise<void> {
		await store.account.cache([
			{
				address: this.senderId,
			},
			{
				username: this.asset.username,
			},
		]);
	}

	protected verifyAgainstTransactions(
		transactions: ReadonlyArray<TransactionJSON>,
	): ReadonlyArray<TransactionError> {
		return transactions
			.filter(
				tx =>
					tx.type === this.type && tx.senderPublicKey === this.senderPublicKey,
			)
			.map(
				tx =>
					new TransactionError(
						'Register delegate only allowed once per account.',
						tx.id,
						'.asset.delegate',
					),
			);
	}

	protected validateAsset(): ReadonlyArray<TransactionError> {
		const schemaErrors = validator.validate(
			delegateAssetFormatSchema,
			this.asset,
		);
		const errors = convertToAssetError(
			this.id,
			schemaErrors,
		) as TransactionError[];

		return errors;
	}

	protected async applyAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const errors: TransactionError[] = [];
		const sender = await store.account.get(this.senderId);

		const usernameExists = store.account.find(
			(account: Account) => account.username === this.asset.username,
		);

		if (usernameExists) {
			errors.push(
				new TransactionError(
					`Username is not unique.`,
					this.id,
					'.asset.username',
				),
			);
		}
		if (sender.isDelegate || sender.username) {
			errors.push(
				new TransactionError(
					'Account is already a delegate',
					this.id,
					'.asset.username',
				),
			);
		}
		const updatedSender = {
			...sender,
			username: this.asset.username,
			vote: 0,
			isDelegate: 1,
		};
		store.account.set(updatedSender.address, updatedSender);

		return errors;
	}

	protected async undoAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const sender = await store.account.get(this.senderId);
		const { username, ...strippedSender } = sender;
		const resetSender = {
			...sender,
			// tslint:disable-next-line no-null-keyword - Exception for compatibility with Core 1.4
			username: null,
			vote: 0,
			isDelegate: 0,
		};
		store.account.set(strippedSender.address, resetSender);

		return [];
	}
}
