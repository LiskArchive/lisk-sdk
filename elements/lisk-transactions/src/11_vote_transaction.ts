/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { stringToBuffer } from '@liskhq/lisk-cryptography';
import { validator } from '@liskhq/lisk-validator';

import {
	BaseTransaction,
	StateStore,
	StateStorePrepare,
} from './base_transaction';
import { VOTE_FEE } from './constants';
import { convertToAssetError, TransactionError } from './errors';
import { TransactionJSON } from './transaction_types';
import { CreateBaseTransactionInput } from './utils';

const PREFIX_UPVOTE = '+';
const PREFIX_UNVOTE = '-';
const MAX_VOTE_PER_ACCOUNT = 101;
const MIN_VOTE_PER_TX = 1;
const MAX_VOTE_PER_TX = 33;

export interface VoteAsset {
	readonly votes: ReadonlyArray<string>;
}

export interface CreateVoteAssetInput {
	readonly unvotes?: ReadonlyArray<string>;
	readonly votes?: ReadonlyArray<string>;
}

export type CastVoteInput = CreateBaseTransactionInput & CreateVoteAssetInput;

export const voteAssetFormatSchema = {
	type: 'object',
	required: ['votes'],
	properties: {
		votes: {
			type: 'array',
			minItems: MIN_VOTE_PER_TX,
			maxItems: MAX_VOTE_PER_TX,
			items: {
				type: 'string',
				format: 'signedPublicKey',
			},
			uniqueSignedPublicKeys: true,
		},
	},
};

interface RawAsset {
	readonly recipientId: string;
	readonly amount: string | number;
	readonly votes: ReadonlyArray<string>;
}

export class VoteTransaction extends BaseTransaction {
	public readonly containsUniqueData: boolean;
	public readonly asset: VoteAsset;
	public static TYPE = 11;
	public static FEE = VOTE_FEE.toString();

	public constructor(rawTransaction: unknown) {
		super(rawTransaction);
		const tx = (typeof rawTransaction === 'object' && rawTransaction !== null
			? rawTransaction
			: {}) as Partial<TransactionJSON>;
		if (tx.asset) {
			const rawAsset = tx.asset as RawAsset;
			this.asset = {
				votes: rawAsset.votes,
			};
		} else {
			// tslint:disable-next-line no-object-literal-type-assertion
			this.asset = {} as VoteAsset;
		}
		this.containsUniqueData = true;
	}

	public assetToJSON(): object {
		return {
			votes: this.asset.votes,
		};
	}

	protected assetToBytes(): Buffer {
		return stringToBuffer(this.asset.votes.join(''));
	}

	public async prepare(store: StateStorePrepare): Promise<void> {
		const publicKeyObjectArray = this.asset.votes.map(pkWithAction => {
			const publicKey = pkWithAction.slice(1);

			return {
				publicKey,
			};
		});
		const filterArray = [
			{
				address: this.senderId,
			},
			...publicKeyObjectArray,
		];

		await store.account.cache(filterArray);
	}

	protected verifyAgainstTransactions(
		transactions: ReadonlyArray<TransactionJSON>,
	): ReadonlyArray<TransactionError> {
		const sameTypeTransactions = transactions
			.filter(
				tx =>
					tx.senderPublicKey === this.senderPublicKey && tx.type === this.type,
			)
			.map(tx => new VoteTransaction(tx));
		const publicKeys = this.asset.votes.map(vote => vote.substring(1));

		return sameTypeTransactions.reduce((previous, tx) => {
			const conflictingVotes = tx.asset.votes
				.map(vote => vote.substring(1))
				.filter(publicKey => publicKeys.includes(publicKey));
			if (conflictingVotes.length > 0) {
				return [
					...previous,
					new TransactionError(
						`Transaction includes conflicting votes: ${conflictingVotes.toString()}`,
						this.id,
						'.asset.votes',
					),
				];
			}

			return previous;
		}, [] as ReadonlyArray<TransactionError>);
	}

	protected validateAsset(): ReadonlyArray<TransactionError> {
		const asset = this.assetToJSON();
		const schemaErrors = validator.validate(voteAssetFormatSchema, asset);
		const errors = convertToAssetError(
			this.id,
			schemaErrors,
		) as TransactionError[];

		return errors;
	}

	protected async applyAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const errors: TransactionError[] = [];
		const sender = await store.account.get(this.senderId);

		this.asset.votes.forEach(actionVotes => {
			const vote = actionVotes.substring(1);
			const voteAccount = store.account.find(
				account => account.publicKey === vote,
			);
			if (
				!voteAccount ||
				(voteAccount &&
					(voteAccount.username === undefined ||
						voteAccount.username === '' ||
						voteAccount.username === null))
			) {
				errors.push(
					new TransactionError(
						`${vote} is not a delegate.`,
						this.id,
						'.asset.votes',
					),
				);
			}
		});
		const senderVotes = sender.votedDelegatesPublicKeys || [];
		this.asset.votes.forEach(vote => {
			const action = vote.charAt(0);
			const publicKey = vote.substring(1);
			// Check duplicate votes
			if (action === PREFIX_UPVOTE && senderVotes.includes(publicKey)) {
				errors.push(
					new TransactionError(
						`${publicKey} is already voted.`,
						this.id,
						'.asset.votes',
					),
				);
				// Check non-existing unvotes
			} else if (action === PREFIX_UNVOTE && !senderVotes.includes(publicKey)) {
				errors.push(
					new TransactionError(
						`${publicKey} is not voted.`,
						this.id,
						'.asset.votes',
					),
				);
			}
		});
		const upvotes = this.asset.votes
			.filter(vote => vote.charAt(0) === PREFIX_UPVOTE)
			.map(vote => vote.substring(1));
		const unvotes = this.asset.votes
			.filter(vote => vote.charAt(0) === PREFIX_UNVOTE)
			.map(vote => vote.substring(1));
		const originalVotes = sender.votedDelegatesPublicKeys || [];
		const votedDelegatesPublicKeys: ReadonlyArray<string> = [
			...originalVotes,
			...upvotes,
		].filter(vote => !unvotes.includes(vote));
		if (votedDelegatesPublicKeys.length > MAX_VOTE_PER_ACCOUNT) {
			errors.push(
				new TransactionError(
					`Vote cannot exceed ${MAX_VOTE_PER_ACCOUNT} but has ${votedDelegatesPublicKeys.length}.`,
					this.id,
					'.asset.votes',
					votedDelegatesPublicKeys.length.toString(),
					MAX_VOTE_PER_ACCOUNT,
				),
			);
		}
		const updatedSender = {
			...sender,
			votedDelegatesPublicKeys,
		};
		store.account.set(updatedSender.address, updatedSender);

		return errors;
	}

	protected async undoAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const errors = [];
		const sender = await store.account.get(this.senderId);

		const upvotes = this.asset.votes
			.filter(vote => vote.charAt(0) === PREFIX_UPVOTE)
			.map(vote => vote.substring(1));
		const unvotes = this.asset.votes
			.filter(vote => vote.charAt(0) === PREFIX_UNVOTE)
			.map(vote => vote.substring(1));
		const originalVotes = sender.votedDelegatesPublicKeys || [];
		const votedDelegatesPublicKeys: ReadonlyArray<string> = [
			...originalVotes,
			...unvotes,
		].filter(vote => !upvotes.includes(vote));
		if (votedDelegatesPublicKeys.length > MAX_VOTE_PER_ACCOUNT) {
			errors.push(
				new TransactionError(
					`Vote cannot exceed ${MAX_VOTE_PER_ACCOUNT} but has ${votedDelegatesPublicKeys.length}.`,
					this.id,
					'.asset.votes',
					votedDelegatesPublicKeys.length.toString(),
					MAX_VOTE_PER_ACCOUNT,
				),
			);
		}

		const updatedSender = {
			...sender,
			votedDelegatesPublicKeys,
		};
		store.account.set(updatedSender.address, updatedSender);

		return errors;
	}
}
