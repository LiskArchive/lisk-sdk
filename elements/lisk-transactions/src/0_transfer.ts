/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { getAddressFromPublicKey } from '@liskhq/lisk-cryptography';
import { TransferTransaction } from './0_transfer_transaction';
import { BYTESIZES, TRANSFER_FEE } from './constants';
import { TransactionJSON } from './transaction_types';
import {
	createBaseTransaction,
	validateAddress,
	validatePublicKey,
	validateTransferAmount,
} from './utils';

export interface TransferInputs {
	readonly amount: string;
	readonly data?: string;
	readonly passphrase?: string;
	readonly recipientId?: string;
	readonly recipientPublicKey?: string;
	readonly secondPassphrase?: string;
	readonly timeOffset?: number;
}

const validateInputs = ({
	amount,
	recipientId,
	recipientPublicKey,
	data,
}: TransferInputs): void => {
	if (!validateTransferAmount(amount)) {
		throw new Error('Amount must be a valid number in string format.');
	}

	if (!recipientId && !recipientPublicKey) {
		throw new Error(
			'Either recipientId or recipientPublicKey must be provided.',
		);
	}

	if (typeof recipientId !== 'undefined') {
		validateAddress(recipientId);
	}

	if (typeof recipientPublicKey !== 'undefined') {
		validatePublicKey(recipientPublicKey);
	}

	if (
		recipientId &&
		recipientPublicKey &&
		recipientId !== getAddressFromPublicKey(recipientPublicKey)
	) {
		throw new Error('recipientId does not match recipientPublicKey.');
	}

	if (data && data.length > 0) {
		if (typeof data !== 'string') {
			throw new Error(
				'Invalid encoding in transaction data. Data must be utf-8 encoded string.',
			);
		}
		if (data.length > BYTESIZES.DATA) {
			throw new Error('Transaction data field cannot exceed 64 bytes.');
		}
	}
};

export const transfer = (inputs: TransferInputs): Partial<TransactionJSON> => {
	validateInputs(inputs);
	const {
		data,
		amount,
		recipientPublicKey,
		passphrase,
		secondPassphrase,
	} = inputs;

	const recipientIdFromPublicKey = recipientPublicKey
		? getAddressFromPublicKey(recipientPublicKey)
		: undefined;
	const recipientId = inputs.recipientId
		? inputs.recipientId
		: recipientIdFromPublicKey;

	const transaction = {
		...createBaseTransaction(inputs),
		asset: data ? { data } : {},
		amount,
		fee: TRANSFER_FEE.toString(),
		recipientId: recipientId as string,
		recipientPublicKey,
		type: 0,
	};

	if (!passphrase) {
		return transaction;
	}

	const transactionWithSenderInfo = {
		...transaction,
		recipientId: recipientId as string,
		senderId: transaction.senderId as string,
		senderPublicKey: transaction.senderPublicKey as string,
	};

	const transferTransaction = new TransferTransaction(
		transactionWithSenderInfo,
	);
	transferTransaction.sign(passphrase, secondPassphrase);

	return transferTransaction.toJSON();
};
