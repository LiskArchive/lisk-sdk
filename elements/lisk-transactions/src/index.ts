/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { DelegateTransaction } from './10_delegate_transaction';
import { MultisignatureTransaction } from './12_multisignature_transaction';
import { VoteTransaction } from './13_vote_transaction';
import { UnlockTransaction } from './14_unlock_transaction';
import { TransferTransaction } from './8_transfer_transaction';
import {
	BaseTransaction,
	StateStore,
	StateStorePrepare,
} from './base_transaction';
import { castVotes } from './cast_votes';
import * as constants from './constants';
import {
	convertToAssetError,
	convertToTransactionError,
	TransactionError,
} from './errors';
import { registerDelegate } from './register_delegate';
import { registerMultisignature } from './register_multisignature_account';
import { createResponse, Status, TransactionResponse } from './response';
import { transactionInterface } from './schema';
import { signMultiSignatureTransaction } from './sign_multi_signature_transaction';
import { Account, TransactionJSON } from './transaction_types';
import { transfer } from './transfer';
import { unlockToken } from './unlock_token';
import {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	getId,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateMultisignatures,
	validateSenderIdAndPublicKey,
	validateSignature,
	verifyMinRemainingBalance,
	verifyMultiSignatureTransaction,
	verifySenderPublicKey,
} from './utils';

const exposedUtils = {
	convertBeddowsToLSK,
	getId,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateMultisignatures,
	validateSignature,
	verifyMinRemainingBalance,
	validateSenderIdAndPublicKey,
	verifyMultiSignatureTransaction,
	verifySenderPublicKey,
};

export {
	Account,
	BaseTransaction,
	StateStore,
	StateStorePrepare,
	TransferTransaction,
	transfer,
	DelegateTransaction,
	registerDelegate,
	VoteTransaction,
	castVotes,
	MultisignatureTransaction,
	UnlockTransaction,
	unlockToken,
	createResponse,
	registerMultisignature,
	signMultiSignatureTransaction,
	Status,
	TransactionResponse,
	TransactionJSON,
	TransactionError,
	transactionInterface,
	convertToAssetError,
	convertToTransactionError,
	constants,
	exposedUtils as utils,
};
