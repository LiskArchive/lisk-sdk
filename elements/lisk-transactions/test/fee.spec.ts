/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { utils, address } from '@liskhq/lisk-cryptography';
import { computeMinFee, getBytes } from '../src';

describe('fee', () => {
	const validParamsSchema = {
		$id: '/lisk/transfer-transaction',
		title: 'Transfer transaction params',
		type: 'object',
		required: ['amount', 'recipientAddress', 'data'],
		properties: {
			amount: {
				dataType: 'uint64',
				fieldNumber: 1,
			},
			recipientAddress: {
				dataType: 'bytes',
				fieldNumber: 2,
				minLength: 20,
				maxLength: 20,
			},
			data: {
				dataType: 'string',
				fieldNumber: 3,
				minLength: 0,
				maxLength: 64,
			},
		},
	};
	const passphrase1 = 'trim elegant oven term access apple obtain error grain excite lawn neck';
	const { publicKey: publicKey1 } = address.getAddressAndPublicKeyFromPassphrase(passphrase1);
	const validTransaction = {
		moduleID: utils.intToBuffer(2, 4),
		commandID: utils.intToBuffer(0, 4),
		nonce: BigInt('1'),
		senderPublicKey: publicKey1,
		params: {
			recipientAddress: Buffer.from('3a971fd02b4a07fc20aad1936d3cb1d263b96e0f', 'hex'),
			amount: BigInt('4008489300000000'),
			data: '',
		},
	};
	const baseFees = [
		{
			moduleID: utils.intToBuffer(2, 4),
			commandID: utils.intToBuffer(0, 4),
			baseFee: '10000000',
		},
		{
			moduleID: utils.intToBuffer(5, 4),
			commandID: utils.intToBuffer(0, 4),
			baseFee: '1',
		},
		{
			moduleID: utils.intToBuffer(3, 4),
			commandID: utils.intToBuffer(0, 4),
			baseFee: '1',
		},
	];

	describe('computeMinFee', () => {
		it('should return minimum fee required to send to network', () => {
			// Arrange
			const minFee = computeMinFee(validTransaction, validParamsSchema);

			// Assert
			expect(minFee).not.toBeUndefined();
			expect(minFee).toMatchSnapshot();
		});

		it('should calculate minimum fee for given minFeePerByte', () => {
			// Arrange
			const options = { minFeePerByte: 2000, baseFees, numberOfSignatures: 1 };
			const minFee = computeMinFee(validTransaction, validParamsSchema, options);

			// Assert
			expect(minFee).not.toBeUndefined();
			expect(minFee).toMatchSnapshot();
		});

		it('should calculate minimum fee for transaction from multisignature account', () => {
			// Arrange
			const options = { minFeePerByte: 2000, baseFees, numberOfSignatures: 64 };
			const minFee = computeMinFee(validTransaction, validParamsSchema, options);

			// Assert
			expect(minFee).not.toBeUndefined();
			expect(minFee).toMatchSnapshot();
		});

		it('should calculate minimum fee for transaction from multisignature account which has lower number of signatures than registered public keys', () => {
			// Arrange
			const options = {
				minFeePerByte: 1000,
				baseFees: [],
				numberOfSignatures: 2,
				numberOfEmptySignatures: 3,
			};
			const transaction = {
				...validTransaction,
				signatures: [
					Buffer.alloc(64),
					Buffer.alloc(0),
					Buffer.alloc(0),
					Buffer.alloc(0),
					Buffer.alloc(64),
				],
			};
			const minFee = computeMinFee(transaction, validParamsSchema, options);
			const txBytes = getBytes({ ...transaction, fee: minFee }, validParamsSchema);

			// Assert
			expect(minFee.toString()).toEqual(BigInt(txBytes.length * 1000).toString());
		});

		it('should calculate minimum fee for delegate registration transaction', () => {
			// Arrange
			const delegateRegisterTransaction = {
				...validTransaction,
				moduleID: utils.intToBuffer(5, 4),
				commandID: utils.intToBuffer(0, 4),
				params: { username: 'delegate1' },
			};
			const options = { minFeePerByte: 1000, baseFees, numberOfSignatures: 1 };
			const delegateRegisterParamsSchema = {
				$id: '/lisk/dpos/register',
				type: 'object',
				required: ['username'],
				properties: {
					username: {
						dataType: 'string',
						fieldNumber: 1,
						minLength: 1,
						maxLength: 20,
					},
				},
			};
			const minFee = computeMinFee(
				delegateRegisterTransaction,
				delegateRegisterParamsSchema,
				options,
			);

			// Assert
			expect(minFee).not.toBeUndefined();
			expect(minFee).toMatchSnapshot();
		});
	});
});
