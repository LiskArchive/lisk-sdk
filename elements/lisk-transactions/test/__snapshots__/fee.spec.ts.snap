// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fee computeMinFee should calculate minimum fee for delegate registration transaction 1`] = `131000n`;

exports[`fee computeMinFee should calculate minimum fee for given minFeePerByte 1`] = `306000n`;

exports[`fee computeMinFee should calculate minimum fee for transaction from multisignature account 1`] = `8624000n`;

exports[`fee computeMinFee should return minimum fee required to send to network 1`] = `153000n`;
